src/lib.rs

```diff
@@ -519,6 +550,72 @@ fn parse_token<'a>(bytes: &mut Bytes<'a>) -> Result<&'a str> {
     }
 }
 
+#[inline]
+fn parse_uri<'a>(bytes: &mut Bytes<'a>) -> Result<&'a str> {
+    #[cfg(feature = "nightly")]
+    while bytes.as_ref().len() >= 32 {
+        let advance = match_url_char_32(bytes.as_ref());
+        bytes.advance(advance);
+
+        if advance != 32 {
+            break;
+        }
+    }
+
+    loop {
+        let b = next!(bytes);
+        if b == b' ' {
+            return Ok(Status::Complete(unsafe {
+                // all bytes up till `i` must have been `is_token`.
+                str::from_utf8_unchecked(bytes.slice_skip(1))
+            }));
+        } else if !is_uri_token(b) {
+            return Err(Error::Token);
+        }
+    }
+}
+
+#[cfg(feature = "nightly")]
+#[cfg(all(any(target_arch = "x86", target_arch = "x86_64"), target_feature = "avx2"))]
```

Comment:
What happens if a user has enabled the `nightly` feature, but their compile target doesn't support these? We'd need `match_url_char_32` versions for those too, I think.

Reply:
I thought about this previous, we need to add an `SSE4.2` version here? Is it reasonable?

I think the completed constaints should be:

* `nightly AND (x86 OR x86_64) AND AVX2` => enable AVX2 support;
* `nightly AND (x86 OR x86_64) AND SSE4.2` => enable SSE4.2 support;
* otherwise, use the fallback implementation.

But this will cause a rather lengthy `#[cfg(...)]`, I don't know how to simply it.

Reply:
Yea, its probably reasonable. Not too worried about complicated cfg attributes.